kernel/generic/src/synch/futex.c

```diff
@@ -343,53 +208,44 @@ static futex_t *get_and_cache_futex(uintptr_t phys_addr, uintptr_t uaddr)
 	 */
 	spinlock_lock(&futex_ht_lock);
 
-	ht_link_t *fut_link = hash_table_find(&futex_ht, &phys_addr);
+	ht_link_t *fut_link = hash_table_find(&futex_ht, &paddr);
 
 	if (fut_link) {
 		free(futex);
 		futex = member_to_inst(fut_link, futex_t, ht_link);
 		futex_add_ref(futex);
 	} else {
-		futex_initialize(futex, phys_addr);
+		futex_initialize(futex, paddr);
 		hash_table_insert(&futex_ht, &futex->ht_link);
 	}
 
 	spinlock_unlock(&futex_ht_lock);
 
-	/*
-	 * Cache the link to the futex object for this task.
-	 */
-	futex_ptr_t *fut_ptr = malloc(sizeof(futex_ptr_t));
-	if (!fut_ptr) {
-		spinlock_lock(&futex_ht_lock);
-		futex_release_ref(futex);
-		spinlock_unlock(&futex_ht_lock);
-		return NULL;
-	}
-	cht_link_t *dup_link;
+	return futex;
+}
 
-	fut_ptr->futex = futex;
-	fut_ptr->uaddr = uaddr;
+/** Finds the physical address of the futex variable. */
+static bool find_futex_paddr(uintptr_t uaddr, uintptr_t *paddr)
+{
+	page_table_lock(AS, false);
+	spinlock_lock(&futex_ht_lock);
 
-	cht_read_lock();
+	bool success = false;
 
-	/* Cache the mapping from the virtual address to the futex for this task. */
-	if (cht_insert_unique(&TASK->futexes->ht, &fut_ptr->cht_link, &dup_link)) {
-		spinlock_lock(&TASK->futexes->list_lock);
-		list_append(&fut_ptr->all_link, &TASK->futexes->list);
-		spinlock_unlock(&TASK->futexes->list_lock);
-	} else {
-		/* Another thread of this task beat us to it. Use that mapping instead.*/
-		free(fut_ptr);
-		futex_release_ref_locked(futex);
+	pte_t t;
+	bool found;
 
-		futex_ptr_t *dup = member_to_inst(dup_link, futex_ptr_t, cht_link);
-		futex = dup->futex;
+	found = page_mapping_find(AS, ALIGN_DOWN(uaddr, PAGE_SIZE), true, &t);
```

Comment:
I feel like this needlessly exposes implementation details. Perhaps we could add a function like `page_mapping_get_paddr()`?